openssl/sha.go

```diff
@@ -9,48 +9,122 @@ package openssl
 // #include "goopenssl.h"
 import "C"
 import (
+	"crypto"
 	"errors"
 	"hash"
+	"runtime"
+	"strconv"
 	"unsafe"
 )
 
-// NewSHA1 returns a new SHA1 hash.
-func NewSHA1() hash.Hash {
-	h := new(sha1Hash)
+type evpHash struct {
+	md        *C.EVP_MD
+	ctx       *C.EVP_MD_CTX
+	ctx2      *C.EVP_MD_CTX
+	size      int
+	blockSize int
+}
+
+func newEvpHash(ch crypto.Hash, size, blockSize int) *evpHash {
+	md := cryptoHashToMD(ch)
+	if md == nil {
+		panic("openssl: unsupported hash function: " + strconv.Itoa(int(ch)))
+	}
+	ctx := C.go_openssl_EVP_MD_CTX_new()
+	ctx2 := C.go_openssl_EVP_MD_CTX_new()
```

Comment:
Why do we need both of these here?

Reply:
`ctx2` is used when calling `Sum`, as the [hash.Hash](https://pkg.go.dev/hash#Hash) interface mandates that Sum does not change the internal state of the sha struct, but we are using `EVP_DigestFinal_ex(ctx, ...)` for doing that operation, which does change the internal state of the supplied context.

In order to avoid that, we copy the main context `h.ctx` into `h.ctx2` and call `EVP_DigestFinal_ex` using `h.ctx2`. This process is explained here:

https://github.com/microsoft/go-crypto-openssl/blob/b9261e0b964072f20be71e76f24221c9b66dec8a/openssl/sha.go#L77-L91

We could also just create a new context inside `Sum`, but we would then be allocating a new object on every Sum call instead of reusing the same object again and again.

I will add more inline comments so future readers don't have this same question 😄